File: 108. Convert Sorted Array to Binary Search Tree/108. Convert Sorted Array to Binary Search Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func sortedArrayToBST(nums []int) *TreeNode {
    if len(nums) == 0 {
        return nil
    }
    
    return buildBST(nums, 0, len(nums)-1)
}

func buildBST(nums []int, start, end int) *TreeNode {
    if start > end {
        return nil
    }
    
    mid := start + (end-start)/2
    node := &TreeNode{Val: nums[mid]}
    node.Left = buildBST(nums, start, mid-1)
    node.Right = buildBST(nums, mid+1, end)
    
    return node
}

File: 108. Convert Sorted Array to Binary Search Tree/108. Convert Sorted Array to Binary Search Tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        Self::build_bst(&nums, 0, (nums.len() as i32) - 1)
    }
    
    fn build_bst(nums: &Vec<i32>, start: i32, end: i32) -> Option<Rc<RefCell<TreeNode>>> {
        if start > end {
            return None;
        }
        
        let mid = start + (end - start) / 2;
        let mut node = TreeNode::new(nums[mid as usize]);
        
        node.left = Self::build_bst(nums, start, mid - 1);
        node.right = Self::build_bst(nums, mid + 1, end);
        
        Some(Rc::new(RefCell::new(node)))
    }
}